README
= This Fork
== Added Functionality

== Updated Rakefile to Jeweler format

The original rake file has been converted to the jeweler format. Several of the 
operations in the original rake file are handled by the default by jeweler, and 
this change brings rgl up to date with current gem management practices. 
Some rake tasks have not been ported (converting to tar/zip, counting code lines,
 cvs tasks) as the utility for them seems to have lessened.

== RGL::DirectedAdjacencyGraph Initialization
 * Nested arrays can be used to initialize a graph

     RGL::DirectedAdjacencyGraph[ [1,2], [2,3], [2,4], [4,5] ].edges_to_a.to_s
     #=> "(1-2)(2-3)(2-4)(4-5)"

     

== Mixed in ForforfRglAdjacency Module into RGL::DirectedAdjacencyGraph
 * Ability to alias nils in the graph
 * #alias_nils(nil_term = :zzznull) will replace any nils in the graph with
   a user provided term (default is :zzznull for alphabetical reasons)
 * overrides #hash and #eql? so that #eql? will return true if two graphs
   have identical set of edges.
 * provides an #in_degree method. Its a bit of a hack, as it essentially reverses
   the graph and then calculates #out_degree on the reversed graph. I'm fairly
   certain this results in a correct answer for directed graphs, but not 100%
   positive.
 * #roots method to find vertices with no sources (i.e., no parents or
   equivalently an in degree of 0)
 * #best_top_vertices method to find the vertices with the most descendants.
   Its vertices, not vertex because multiple vertices could be tied for the
   number of descendants.
 * #edge_array returns an array of edges interestingly enough
 * #source_vertices returns the parent vertices of the current vertex
 * #connected_to? determines if the current graph (self) has any vertices in
   common with another graph.
 * #merge will merge the current directed graph with another directed graph.
 * connected_edges? A test to determine if edges connect to each other. For
   Example edge [:a, :b] is connected to [:b, ;c].
 * #atomic_graphs returns an array of digraphs that are formed by the edges of
   the current graph (i.e. each digraph returned has two vertices and one edge).
   I wish I could have thought of a better name (atomic has multiple meanins)
 * #find_connected_graphs will find all connected graphs and merge them together
   into a common graph. Returns an array of unique graphs.

= Ruby Graph Library (RGL)

RGL is a framework for graph data structures and algorithms.

The design of the library is much influenced by the Boost Graph Library (BGL)
which is written in C++ heavily using its template mechanism. Refer to
http://www.boost.org/libs/graph/doc for further links and documentation on graph
data structures and algorithms and the design rationales of BGL.

A comprehensive summary of graph terminology can be found in the the graph
section of the <em>Dictionary of Algorithms and Data Structures</em> at
http://www.nist.gov/dads/HTML/graph.html.

== Design principles

This document concentrates on the special issues of the implementation in
Ruby. The main design goals directly taken from the BGL design are:

* An interface for how the structure of a graph can be accessed using a generic
  interface that hides the details of the graph data structure
  implementation. This interface is defined by the module Graph, which should be
  included in concrete classes.

* A standardized generic interface for traversing graphs (RGL::GraphIterator)

RGL provides some general purpose graph classes that conform to this interface,
but they are not meant to be the *only* graph classes. As in BGL I believe that
the main contribution of the RGL is the formulation of this interface.

The BGL graph interface and graph components are generic in the sense of the C++
Standard Template Library (STL). In Ruby other techniques are available to
express the generic character of the algorithms and data structures mainly using
mixins and iterators. The BGL documentation mentions three means to achieve
genericity:

* Algorithm/Data-Structure Interoperability
* Extension through Function Objects and Visitors
* Element Type Parameterization
* Vertex and Edge Property Multi-Parameterization

The first is easily achieved in RGL using mixins, which of course is not as
efficient than C++ templates (but much more readable :-). The second one is
even more easily implemented using standard iterators with blocks or using the
Stream[http://rgl.rubyforge.org/stream/files/README.html] module. The third one
is no issue since Ruby is dynamically typed: Each object can be a graph
vertex. There is no need for a vertex (or even edge type). In the current
version of RGL properties of vertices are simply attached using hashes. At
first there seems to be not much need for the graph property machinery.

=== Algorithms

The first version of RGL only contains a core set of algorithm patterns:

* Breadth First Search (RGL::BFSIterator)
* Depth First Search (RGL::DFSIterator)

The algorithm patterns by themselves do not compute any meaningful quantities
over graphs, they are merely building blocks for constructing graph
algorithms. The graph algorithms in RGL currently include:

* Topological Sort (RGL::TopsortIterator)
* Connected Components (RGL::Graph#each_connected_component)
* Strongly Connected Components (RGL::Graph#strongly_connected_components)
* Transitive Closure (RGL::Graph#transitive_closure)

=== Data Structures

RGL currently provides two graph classes that implement a generalized adjacency
list and an edge list adaptor.

* RGL::AdjacencyGraph
* RGL::ImplicitGraph

The AdjacencyGraph class is the general purpose *swiss army knife* of graph
classes. It is highly parameterized so that it can be optimized for different
situations: the graph is directed or undirected, allow or disallow parallel
edges, efficient access to just the out-edges, fast vertex insertion and removal
at the cost of extra space overhead, etc.

=== Differences to BGL

The concepts of IncidenceGraph, AdjacencyGraph and VertexListGraph (see
http://www.boost.org/libs/graph/doc/IncidenceGraph.html) are here bundled in the
base graph module. Most methods of IncidenceGraph should be standard in the base
module Graph. The complexity guarantees can not necessarily provided.  See
http://www.boost.org/libs/graph/doc/graph_concepts.html.

== Installation

RGL is depended on the
stream[http://rgl.rubyforge.org/stream/files/README.html] library which can
also be downloaded from http://rubyforge.org/frs/?group_id=110. If you use gem
to install RGL the stream library will be installed as a prerequisite.

=== GEM Installation

Download the GEM file and install it with ..

  % gem install rgl-VERSION.gem

or directly with 

  % gem install rgl

Use the correct version number for VERSION (e.g. 0.2.x).  You may need root
privileges to install.

=== Running tests

RGL comes with a Rakefile which automatically runs the tests. Goto the
installation directory and start rake:

 % gem env
 Rubygems Environment:
  - VERSION: 0.9.0 (0.9.0)
  - INSTALLATION DIRECTORY: /usr/lib/ruby/gems/1.8
  - GEM PATH:
     - /usr/lib/ruby/gems/1.8
  - REMOTE SOURCES:
     - http://gems.rubyforge.org

 % cd /usr/lib/ruby/gems/1.8/gems/rgl-0.3.0/
 % rake
 (in /usr/lib/ruby/gems/1.8/gems/rgl-0.3.0)
 /usr/bin/ruby1.8 -Ilib:tests "/usr/lib/ruby/gems/1.8/gems/rake-0.7.3/lib/rake/rake_test_loader.rb" "tests/TestTransitiveClosure.rb" "tests/TestComponents.rb" "tests/TestCycles.rb" "tests/TestDirectedGraph.rb" "tests/TestEdge.rb" "tests/TestGraph.rb" "tests/TestGraphXML.rb" "tests/TestImplicit.rb" "tests/TestUnDirectedGraph.rb" "tests/TestTraversal.rb" "tests/TestDot.rb" "tests/TestRdot.rb" 
 Loaded suite /usr/lib/ruby/gems/1.8/gems/rake-0.7.3/lib/rake/rake_test_loader
 Started
 ......................................................................................................................................................
 Finished in 0.750958 seconds.

 86 tests, 625 assertions, 0 failures, 0 errors
 
=== Code coverage

Running rcov[http://eigenclass.org/hiki.rb?rcov] on the test suite generates this[link:coverage/index.html] result.

=== Normal Installation

You have to install stream library before. You can than install RGL with the
following command:

  % ruby install.rb

from its distribution directory. To uninstall it use

  % ruby install.rb -u

== Example irb session with RGL

  irb> require 'rgl/adjacency'
  irb> dg=RGL::DirectedAdjacencyGraph[1,2 ,2,3 ,2,4, 4,5, 6,4, 1,6]
  # Use DOT to visualize this graph:
  irb> require 'rgl/dot'
  irb> dg.write_to_graphic_file('jpg')
  "graph.jpg"
  
The result: link:../examples/example.jpg  
  
  irb> dg.directed?
  true
  irb> dg.vertices
  [5, 6, 1, 2, 3, 4]
  irb> dg.has_vertex? 4
  true

Every object could be a vertex (there is no class Vertex), even the class
object _Object_:

  irb> dg.has_vertex? Object
  false
  irb> dg.edges.sort.to_s
  "(1-2)(1-6)(2-3)(2-4)(4-5)(6-4)"
  irb> dg.to_undirected.edges.sort.to_s
  "(1=2)(1=6)(2=3)(2=4)(5=4)(6=4)"

Add inverse edge (4-2) to directed graph:

  irb> dg.add_edge 4,2

(4-2) == (2-4) in the undirected graph:

  irb> dg.to_undirected.edges.sort.to_s
  "(1=2)(1=6)(2=3)(2=4)(5=4)(6=4)"

(4-2) != (2-4) in directed graphs:

  irb> dg.edges.sort.to_s
  "(1-2)(1-6)(2-3)(2-4)(4-2)(4-5)(6-4)"
  irb> dg.remove_edge 4,2
  true

<em>Topological sort</em> is realized with as iterator:

  require 'rgl/topsort'
  irb> dg.topsort_iterator.to_a
  [1, 2, 3, 6, 4, 5]

A more elaborated example showing <em>implicit graphs</em>:

  def module_graph
	RGL::ImplicitGraph.new { |g|
	  g.vertex_iterator { |b|
		ObjectSpace.each_object(Module, &b) 
	  }
	  g.adjacent_iterator { |x, b|
		x.ancestors.each { |y|
		  b.call(y) unless x == y || y == Kernel || y == Object
		}
	  }
	  g.directed = true
	}
  end

This function creates a directed graph, with vertices being all loaded modules:

  g = module_graph

We only want to see the ancestors of RGL::AdjacencyGraph:

  tree = bfs_search_tree_from(g,RGL::AdjacencyGraph)

Now we want to visualize this component of g with DOT.  We therefore create a
subgraph of the original graph, using a filtered graph:

  g = g.vertices_filtered_by {|v| tree.has_vertex? v}

Create the graphics with DOT:

  g.write_to_graphic_file

produces module_graph.jpg: link:../examples/module_graph.jpg

Look for more in the examples directory (i.e.
examples.rb[link:files/examples/examples_rb.html]).

== My del.icio.us links concerning RGL

I collect some links to stuff around RGL at http://del.icio.us/monora/rgl. I
registered RGL at SWiK[http://swik.net/rgl].

== Credits

Many thanks to Robert Feldt which also worked on a graph library
(http://rockit.sf.net/subprojects/graphr) who pointed me to BGL and many other
graph resources.

Robert kindly allowed to integrate his work on graphr, which I did not yet
succeed. Especially his work to output graphs for
GraphViz[http://www.research.att.com/sw/tools/graphviz/download.html] is much
more elaborated than the minimal support in dot.rb.

Jeremy Siek one of the authors of the nice book "The Boost Graph Library (BGL)"
(http://www.boost.org/libs/graph/doc) kindly allowed to use the
BGL documentation as a _cheap_ reference for RGL. He and Robert also gave
feedback and many ideas for RGL.

Dave Thomas for RDoc[http://rdoc.sourceforge.net] which generated what you read
and matz for Ruby. Dave included in the latest version of RDoc (alpha9) the
module dot/dot.rb which I use instead of Roberts module to visualize graphs
(see rgl/dot.rb).

Jeremy Bopp, John Carter, Sascha Doerdelmann and Shawn Garbett for contributing
additions, test cases and bugfixes.

== Copying

RGL is Copyright (c) 2002,2004,2005,2008 by Horst Duchene.  It is free software, and may be
redistributed under the terms specified in the README file of the Ruby distribution.

== Support

Please contact me at mailto:monora@gmail.com with bug reports
suggestions, and other comments. If you send patches, it would help if
they were in-line (not attachments) and generated using "diff -u".




ChangeLog
2011-04-15 19:59  forforf

  * ChangeLog: Using git to manage changes, change log will not be updated

2008-08-28 01:36  javanthropus

	* ChangeLog (tags: REL_0_4_0): pre-tag commit

2008-08-28 01:36  javanthropus

	* Rakefile (tags: REL_0_4_0): Update rdoc2rf task to use rsync and
	  more generic remote path specification

2008-08-27 23:30  javanthropus

	* lib/rgl/base.rb (tags: REL_0_4_0): Preparing for 0.4.0 release

2008-08-26 20:07  javanthropus

	* lib/rgl/dot.rb, lib/rgl/rdot.rb, tests/TestRdot.rb: Move the DOT
	  module into the RGL module

	  * This eliminates a class conflict with the DOT module from rdoc
	  when building RGL's documentation * Also remove the superfluous
	  DOT prefixes from class names in the DOT module

2008-08-24 06:16  javanthropus

	* Rakefile: Remove some comments I accidentally left in while
	  testing rdoc functionality

2008-08-24 06:03  javanthropus

	* Rakefile, lib/rgl/transitiv_closure.rb, lib/rgl/transitivity.rb,
	  tests/TestTransitiveClosure.rb, tests/TestTransitivity.rb:
	  Feature 21641: Added transitive reduction functionality

	  * Updated the gem description to announce this functionality *
	  Moved the transitive closure functionality into the
	  transitivity.rb file along with the transitive reduction
	  funtionality * Modifed the transitiv_closure.rb file to simply
	  load the transitivity.rb file for backward compatibility * Moved
	  all transitivity tests into TestTransitivity.rb

2008-08-23 15:45  javanthropus

	* lib/rgl/condensation.rb, lib/rgl/transitiv_closure.rb,
	  tests/TestTransitiveClosure.rb: Defect 21630: Fixed transitive
	  closure

	  * The fix is based on the algorithm described in the
	  documentation for the implementation of transitive closure in
	  Boost * Along with the fix, performance is improved to O(|V||E|)
	  * This implementation needs graph condensation, so that function
	  was added as well * More tests were added to cover more corner
	  cases

2008-08-23 05:40  javanthropus

	* tests/TestGraph.rb: Update basic graph tests to account for
	  graphs with edgeless vertices Also clean up some minor formatting
	  and assertion issues

2008-08-23 05:37  javanthropus

	* lib/rgl/adjacency.rb: Defect 21609: Fix the to_adjacency method
	  to preserve edgeless vertices

2008-03-18 15:03  javanthropus

	* lib/rgl/rdot.rb, tests/TestRdot.rb: More reliably detect and
	  handle newlines embedded within IDs and labels

2008-03-08 10:48  monora

	* ChangeLog, lib/rgl/base.rb (utags: REL_0_3_1): Prepare 0.3.1
	  release

2008-03-04 20:18  monora

	* Rakefile (tags: REL_0_3_1, REL_0_3_0): pre-tag commit

2008-03-02 18:16  javanthropus

	* lib/rgl/rdot.rb (tags: REL_0_3_0): IDs and labels must be
	  converted to strings before processing

2008-03-02 17:45  javanthropus

	* lib/rgl/rdot.rb, tests/TestRdot.rb: Added documentation for
	  rdot.rb and full test coverage

2008-03-02 15:19  monora

	* README: Removed dead link to rubygarden.com

2008-03-02 15:09  monora

	* ChangeLog, README, Rakefile (utags: REL_0_3_0): Polishing before
	  0.3.0 release

2008-03-02 13:45  monora

	* lib/rgl/: adjacency.rb, rdot.rb (utags: REL_0_3_0): Fixed
	  warnings generated by Ruby Dev Tools

2008-03-02 07:57  javanthropus

	* Rakefile, lib/rgl/graphxml.rb (tags: REL_0_3_0),
	  tests/TestGraphXML.rb (tags: REL_0_3_0): Change the way GraphML
	  support is added to the MutableGraph module such that the
	  interface is cleaner and documented

2008-03-01 20:12  javanthropus

	* lib/rgl/base.rb, tests/TestGraph.rb: Polish the documentation for
	  and expand the test coverage of RGL::Graph

2008-02-27 19:44  monora

	* README: Added link to coverage page

2008-02-26 06:01  javanthropus

	* lib/rgl/dot.rb, lib/rgl/rdot.rb, tests/TestDot.rb,
	  tests/TestRdot.rb: All IDs for DOT objects, including names,
	  options, and values, are now automatically quoted as necessary
	  according to the rules documented at
	  http://www.graphviz.org/doc/info/lang.html and
	  http://www.graphviz.org/Documentation/dotguide.pdf.

	  Labels are handled specially in order to account for \l, \r, and
	  \n sequences but are otherwise treated the same as other options.

	  New tests were added to confirm proper function.

	  Some changes were made in order to remove explicit quotes from
	  labels which are no longer necessary.

2008-02-17 20:15  monora

	* lib/rgl/enumerable_ext.rb: Removed backwards compatability method
	  inject. Ruby > 1.8 supports it out of the box.

2008-02-17 20:07  monora

	* tests/: TestEdge.rb, TestDirectedGraph.rb, TestGraph.rb: Improved
	  test coverage

2008-02-17 20:06  monora

	* lib/rgl/adjacency.rb: fixed bug in edgelist_class

2008-02-17 17:59  monora

	* Rakefile: - added coverage task - fixed BUG #2674 - added
	  changelog task - use jamis buck rdoc template

2008-02-17 17:45  monora

	* lib/rgl/base.rb: Changed Version to 0.3.0

2008-02-17 09:09  javanthropus

	* lib/rgl/rdot.rb: DOTSimpleElement provides no useful function, so
	  remove it

2008-02-17 09:08  javanthropus

	* tests/TestRdot.rb: Test that setting only the name for a DOTNode
	  does NOT set the label

2008-02-17 08:58  javanthropus

	* lib/rgl/rdot.rb, tests/TestRdot.rb: Add support for the Mrecord
	  shape to DOTNode.  Rewrite DOTNode#to_s to be easier to
	  understand.  #Rewrite DOTPort to allow for nesting ports.

2008-02-17 03:20  javanthropus

	* lib/rgl/rdot.rb, tests/TestRdot.rb: BUG 17964: DOTElement no
	  longer sets the label unless the user explicitly sets one

2008-02-17 02:56  javanthropus

	* tests/TestRdot.rb: Fix a DOTEdge test which was actually
	  retesting DOTNode

2008-02-16 19:58  javanthropus

	* lib/rgl/rdot.rb, tests/TestRdot.rb: BUG #17962: Subgraphs must be
	  identified by a "subgraph" header rather than a "graph" header

2008-02-13 22:32  monora

	* tests/TestRdot.rb: Fixed typo

2008-02-13 22:20  monora

	* doc/jamis.rb: Added template from Jamis Buck for rdoc. Looks
	  better.

2008-02-12 23:37  monora


spec/spec_helper.rb
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.dirname(__FILE__))
require 'rspec'
require 'rgl/adjacency'

# Requires supporting files with custom matchers and macros, etc,
# in ./support/ and its subdirectories.
Dir["#{File.dirname(__FILE__)}/support/**/*.rb"].each {|f| require f}

RSpec.configure do |config|
  
end


spec/forforf_rgl_adjacency_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/spec_helper')


MyDG = RGL::DirectedAdjacencyGraph


module DAGSpecH

 
  ComplexGraph = [ [:aa, :a], [:bc, :bbb], [:bc,:b], [:aaa, :aa], [:b, nil], [:ba,:b],
                 [:ba, :ab], [:ac, :a], [:bbb, :bb], [:bbb, :aaa], [:a, :aa],
                 [:ab, :a], [:ab, :aaa], [:ab, :bb], [:bcc, :bc], [:bb, :b],
                 #this edge is unconnected to the rest of the graph
                 [:cc, :c], [nil, nil], [nil, :d], [:d, nil ] ]
   CGInDegree = { :a   => 3, :b   => 3, :c   => 1,
                  :aa  => 2, :ab  => 1, :ac  => 0,
                  :aaa => 2, 
                  :ba  => 0, :bb  => 2, :bc  => 1,
                  :bbb => 1, :bcc => 0,
                  :cc  => 0,
                  :d => 1,
                  nil => 3 } 

   CGOutDegree = { :a   => 1, :b   => 1, :c   => 0,
                   :aa  => 1, :ab  => 3, :ac  => 1,
                   :aaa => 1,
                   :ba  => 2, :bb  => 1, :bc  => 2,
                   :bbb => 2, :bcc => 1,
                   :cc => 1,
                   :d => 1,
                   nil => 2 } 

   CGTopNodes = [:ba, :ac, :bcc, :cc] 
   CGBestTopNodes = [:bcc] #underneath :bcc => :bc, :bbb, :aaa, :aa, :a, :bb, :b, nil, :d 
   SimpleGraph = [ [:a, :aa], [:a, :ab], [:b, :ba], [:b, :bb] ]
   SimpleGraphCopy = [ [:a, :aa], [:a, :ab], [:b, :ba], [:b, :bb] ]
   SimpleGraphReverse = [ [:aa, :a], [:ab, :a], [:ba, :b], [:bb, :b] ]
   ConnectRootsGraph = [[:e, :bcc], [:e, :cc], [:e, :ba], [:e, :ac] ]
   ConnectLoopGraph = [[:d, :f], [:a, :f], [:f, :bcc] ]
   UnconnectedGraph = [[:x, :xx], [:z, :zz]]
   NilGraph = [[nil, nil]]
end

describe "ForforfRglAdjacency" do
  include DAGSpecH

  before(:each) do
    #rgl formats shouldn't be needed if using forked rgl

    @complex = ComplexGraph
    @rgl_complex = @complex.flatten
    @complex_indegree = CGInDegree
    @complex_outdegree =CGOutDegree
    @complex_top_nodes = CGTopNodes
    @complex_best_top_nodes = CGBestTopNodes
    @simple = SimpleGraph
    @rgl_simple = @simple.flatten
    @simple_copy = SimpleGraphCopy
    @rgl_simple_copy = @simple_copy.flatten
    @simple_reverse = SimpleGraphReverse
    @rgl_simple_reverse = @simple_reverse.flatten
    @connect_roots = ConnectRootsGraph
    @rgl_connect_roots = @connect_roots.flatten
    @connect_loop = ConnectLoopGraph
    @rgl_connect_loop = @connect_loop.flatten
    @unconnected_graph = UnconnectedGraph
    @rgl_unconnected_graph = @unconnected_graph.flatten
    @nil_graph = NilGraph
    @rgl_nil_graph = @nil_graph.flatten
  end

  it "should initialize" do
    dg_old_way = MyDG[*@rgl_complex]
    dg_old_way.vertices.size.should == 15
    dg_nested_array = MyDG[*@complex]
    dg_nested_array.vertices.size.should == 15
  end

  it "should be able to alias nils" do
    dg_w_nils = MyDG[*@rgl_complex]
    dg_w_nils.edge_array.flatten.should include nil
    dg = dg_w_nils.alias_nils
    dg.edge_array.flatten.should_not include nil
  end

  it "tests for equality" do
    dg_orig = MyDG[*@rgl_simple]
    dg_copy = MyDG[*@rgl_simple_copy]
    dg_orig.hash.should == dg_copy.hash
    same = dg_orig.eql?(dg_copy)
    same.should == true
    dg_diff = MyDG[*@rgl_simple_reverse]
    diff = dg_diff.eql?(dg_orig)
    diff.should == false
  end

  it "counts in-degree" do
    dg = MyDG[*@rgl_complex]
    in_degrees = {}
    @rgl_complex.each do |vert|
      in_degrees[vert] = dg.in_degree(vert)
    end
    in_degrees.should == @complex_indegree 
  end

  it "counts out-degree" do
    dg = MyDG[*@rgl_complex]
    out_degrees = {}
    @rgl_complex.each do |vert|
      out_degrees[vert] = dg.out_degree(vert)
    end
    out_degrees.should == @complex_outdegree
  end

  it "find root vertices (vertices with no parents)" do
    dg = MyDG[*@rgl_complex]
    dg.roots.should == @complex_top_nodes
  end

  it "finds the best top vertices (vertices with the most total descendants)" do
    dg = MyDG[*@rgl_complex]
    dg.best_top_vertices.should == @complex_best_top_nodes
  end

  it "provides its edges as a nested array" do
    dg = MyDG[*@rgl_complex]
    #nils in arrays are kind of a pain
    dg_edges = dg.edge_array.flatten
    complex_edges = @complex.flatten
    #replace nils with :znull
    bye_nil = lambda{|e| e || :znull}
    dg_edges.map! &bye_nil
    complex_edges.map! &bye_nil
    dg_edges.sort.should == complex_edges.sort
  end

  it "finds the source (parent) vertices" do
    dg = MyDG[*@rgl_complex]
    dg.source_vertices(:b).sort.should == [:bc, :ba, :bb].sort
    dg.source_vertices(nil).should == [:b, nil, :d]
    dg.source_vertices(:bcc).should == []
    dg.source_vertices(:a).sort.should == [:aa, :ac, :ab].sort
    dg.source_vertices(:d).should == [nil]
  end

  it "determines if digraphs are connected" do
    base_dg = MyDG[*@rgl_complex]
    conn_root_dg = MyDG[*@rgl_connect_roots]
    unconn_dg = MyDG[*@rgl_unconnected_graph]
    nil_dg = MyDG[*@rgl_nil_graph]

    base_dg.connected_to?(conn_root_dg).should == true
    base_dg.connected_to?(unconn_dg).should == false
    base_dg.connected_to?(nil_dg).should == true 
  end

  it "merges with other graphs" do
    base_dg = MyDG[*@rgl_complex]
    conn_root_dg = MyDG[*@rgl_connect_roots]
    conn_loop_dg = MyDG[*@rgl_connect_loop]
    unconn_dg = MyDG[*@rgl_unconnected_graph]
    
    merge_new_root_dg = base_dg.merge(conn_root_dg)
    merge_new_root_dg.roots.should == [:e]

    merge_new_loop_dg = base_dg.merge(conn_loop_dg)
    merge_new_loop_dg.adjacent_vertices(:d).should include :f
    merge_new_loop_dg.adjacent_vertices(:f).should include :bcc
    merge_new_loop_dg.adjacent_vertices(:a).should include :f
    
    merge_uncon_dg = base_dg.merge(unconn_dg)
    merge_uncon_dg.vertices.should include :bb
    merge_uncon_dg.vertices.should include :cc
    merge_uncon_dg.vertices.should include :x
  end

  it "detects if two edges are connected" do
    base_edge = [:g, :h]
    conn_source = [:g, :i]
    conn_target = [:j, :h]
    linked_edge = [:h, :n]
    dup_edge = [:g, :h]
    unconn_edge = [:k, :m]
    nil_base = [nil, nil]
    nil_conn = [:o, nil]
    
    some_dg = MyDG[*[]] #this method doesn't rely on any dg structures
    some_dg.connected_edges?(base_edge, conn_source).should == [:g, :h, :i]
    some_dg.connected_edges?(base_edge, conn_target).should == [:g, :h, :j]
    some_dg.connected_edges?(base_edge, linked_edge).should == [:g, :h, :n]
    some_dg.connected_edges?(base_edge, dup_edge).should == [:g, :h]
    some_dg.connected_edges?(base_edge, unconn_edge).should == nil
    some_dg.connected_edges?(nil_base, nil_conn).should == [nil, :o]
  end

  it "breaks a graph down into its atomic digraphs" do
    base_dg = MyDG[*@rgl_simple]
    baby_dgs = base_dg.atomic_graphs

    baby_edge_matcher = [[:a, :aa], [:a, :ab], [:b, :ba], [:b, :bb]]
    baby_dgs.size.should == 4
    baby_dgs.each do |dg|
      dg.size.should == 2
      dg_edge = dg.edge_array.flatten #or could have used #first
      baby_edge_matcher.should include dg_edge
      baby_edge_matcher.delete(dg_edge) #makes sure there are no duplicates
    end
  end

  it "can find connected graphs" do
    #simple default test
    base_dg = MyDG[*@rgl_simple]
    base_baby_graphs = base_dg.find_connected_graphs
    base_baby_graphs.size.should == 2
    new_graph = base_baby_graphs.first.merge(base_baby_graphs.last)
    new_graph.should == base_dg

    #complex default test
    complex_dg = MyDG[*@rgl_complex] 
    complex_baby_graphs = complex_dg.find_connected_graphs

    new_complex_graph = complex_baby_graphs.inject(MyDG[*[]]) {|memo, dg| memo.merge(dg)}
    new_complex_graph.should == complex_dg

    #non-default test
    simp_dg = MyDG[*@rgl_simple]
    baby_dgs = simp_dg.atomic_graphs
    baby_dgs.size.should == 4

    some_dg = MyDG[*[]]
    merged_graphs = some_dg.find_connected_graphs(baby_dgs)
    merged_graphs.size.should == 2
    
    full_graph = merged_graphs.first.merge(merged_graphs.last)
    full_graph.should == simp_dg 
  end
end


examples/insel-der-tausend-gefahren.rb
# Die Insel der 1000 Gefahren
# http://www.amazon.de/1000-Gefahren-Die-Insel/dp/3473520225/ref=pd_sim_b?ie=UTF8&qid=1203279845&sr=8-3

require 'rgl/adjacency'
require 'rgl/implicit'
require 'rgl/dot'

g = RGL::DirectedAdjacencyGraph[
  8, 9,
  8, 10,

  9, 11,
  9, 12,

  10, 13,
  10, 14,

  11, 15,
  11, 16,
 
  12, 17,
  12, 18, 

  13, 19, 
  13, 20,

  14, 21,
  14, 22,
 
  15, 23,
  15, 24,

  16, 25,
  16, 26,

  17, 27,
  17, 28,

  18, 29,
  18, 30, 

  18, 31,
  18, 32,

  19, 31, 
  19, 32,

  20, 33,
  20, 34,

  21, 35,
  21, 36,

  22, 37,
  22, 38,

  23, 39,
  23, 40,

  24, 41, 
  24, 42,

  25, 43,
  25, 44,

  26, 45,
  26, 46,

  27, 47,
  27, 48,

  28, 49,
  28, 50, 

  29, 51,
  29, 52,  

  30, 53,
  30, 54,
  
  31, 55,
  31, 56, 
  
  32, 57,
  32, 58,

  33, 59,
  33, 60,
 
  34, 61,
  34, 62,

  35, 63,
  35, 64,
#
  36, 65,
  36, 66,

  37, 67,

  38, 13,

  39, 68,
  39, 69,

  40, 70,
  40, 71,

  42, 72,
  42, 73,

  43, 74,
  43, 75,

  44, 76,
  44, 77,

  46, 78,
  46, 79,

  47, 80,
  47, 81,

  48, 82,
  48, 83,

  50, 84,
  50, 85,

  51, 86,
  51, 87,

  53, 90,
  53, 91,

  55, 93,
  55, 94, 

  
  57,  92,


  58, 97,
  58, 98,

  59, 99,
  99, 100,

  60, 100,


  61, 101,
  61, 102,
  61, 103,

  62, 104,
  62, 105, 

  64, 88,

  65, 106,
  65, 107,

  66, 9,

  67, 108,
  67, 109,

  75, 9,
  
  78, 26,

  86, 30,

  88, 106,
  88, 89,

  89, 16,

  92, 95,
  92, 96,

  93, 37,

  96, 9,

  105, 9,
  105, 10]
g.dotty
